thread-manager/src/native_thread_runtime.rs

```diff
@@ -0,0 +1,153 @@
+use {
+    crate::policy::{apply_policy, parse_policy, CoreAllocation},
+    anyhow::bail,
+    log::error,
+    serde::{Deserialize, Serialize},
+    solana_metrics::datapoint_info,
+    std::{
+        ops::Deref,
+        sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc, Mutex,
+        },
+    },
+};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(default)]
+pub struct NativeConfig {
+    pub core_allocation: CoreAllocation,
+    pub max_threads: usize,
+    /// Priority in range 1..99
+    pub priority: u8,
+    pub policy: String,
+    pub stack_size_bytes: usize,
+}
+
+impl Default for NativeConfig {
+    fn default() -> Self {
+        Self {
+            core_allocation: CoreAllocation::OsDefault,
+            max_threads: 16,
+            priority: crate::policy::DEFAULT_PRIORITY,
+            policy: "OTHER".to_owned(),
+            stack_size_bytes: 2 * 1024 * 1024,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct NativeThreadRuntimeInner {
+    pub id_count: AtomicUsize,
+    pub running_count: Arc<AtomicUsize>,
+    pub config: NativeConfig,
+    pub name: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct NativeThreadRuntime {
+    inner: Arc<NativeThreadRuntimeInner>,
+}
+
+impl Deref for NativeThreadRuntime {
+    type Target = NativeThreadRuntimeInner;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+pub struct JoinHandle<T> {
+    std_handle: Option<std::thread::JoinHandle<T>>,
+    running_count: Arc<AtomicUsize>,
+}
+
+impl<T> JoinHandle<T> {
+    fn join_inner(&mut self) -> std::thread::Result<T> {
+        match self.std_handle.take() {
+            Some(jh) => {
+                let result = jh.join();
+                let rc = self.running_count.fetch_sub(1, Ordering::Relaxed);
+                datapoint_info!("thread-manager-native", ("threads-running", rc, i64),);
+                result
+            }
+            None => {
+                panic!("Thread already joined");
+            }
+        }
+    }
+
+    pub fn join(mut self) -> std::thread::Result<T> {
+        self.join_inner()
+    }
+
+    pub fn is_finished(&self) -> bool {
+        match self.std_handle {
+            Some(ref jh) => jh.is_finished(),
+            None => true,
+        }
+    }
+}
+
+impl<T> Drop for JoinHandle<T> {
+    fn drop(&mut self) {
+        if self.std_handle.is_some() {
+            error!("Attempting to drop a Join Handle of a running thread will leak thread IDs, please join your managed threads!");
+            self.join_inner().expect("Child thread panicked");
+        }
+    }
+}
+
+impl NativeThreadRuntime {
+    pub fn new(name: String, cfg: NativeConfig) -> Self {
```

Comment:
we have a soft rule that thread names should not exceed 15 characters; wdyt of adding some check here to make sure the `name` is <12 characters (so that it can have the -{:2} suffix)?

Reply:
Yes, good idea. Done across all runtimes.